### PR TITLE
Dynamic report severity list

### PR DESCRIPTION
**Pull request by lilian-houdelet**

### CODE REVIEW

File: src/main/java/org/gridsuite/report/server/ReportService.java [hunk 1]
```diff
@@ -113,6 +114,13 @@ public ReporterModel getReport(UUID reportId, boolean withElements, Set<String>
         return report;
     }
 
+    @Transactional(readOnly = true)
+    public List<String> getReportSeverity(UUID reportId, boolean withElements, String reportNameFilter, ReportNameMatchingType reportNameMatchingType) {
+
+        var report = getReport(reportId, withElements, Stream.of(SeverityLevel.values()).map(Enum::name).collect(Collectors.toSet()), reportNameFilter, reportNameMatchingType);
```
Comment (flomillot) [inline src/main/java/org/gridsuite/report/server/ReportService.java:120]: Can you extract the stream ? It's not easily readable

File: src/main/java/org/gridsuite/report/server/ReportService.java [hunk 2]
```diff
@@ -123,6 +131,25 @@ public ReporterModel getSubReport(UUID reporterId, Set<String> severityLevels) {
         return report;
     }
 
+    @Transactional(readOnly = true)
+    public List<String> getSubReportSeverity(UUID reporterId) {
+        var report = getSubReport(reporterId, Stream.of(SeverityLevel.values()).map(Enum::name).collect(Collectors.toSet()));
```
Comment (flomillot) [inline src/main/java/org/gridsuite/report/server/ReportService.java:136]: Same

File: src/main/java/org/gridsuite/report/server/ReportService.java [hunk 3]
```diff
@@ -123,6 +131,25 @@ public ReporterModel getSubReport(UUID reporterId, Set<String> severityLevels) {
...
+    private List<String> getNotificationSecurityLevelList(ReporterModel reporterModel) {
+        if (reporterModel.getSubReporters().isEmpty()) {
+            return reporterModel.getReports().stream().map(report -> report.getValue("reportSeverity").getValue().toString()).distinct().toList();
+        }
+        List<String> reportSeverityList = new ArrayList<>(reporterModel.getSubReporters().stream()
+                .map(this::getNotificationSecurityLevelList)
+                .flatMap(Collection::stream)
+                .toList());
+
+        reportSeverityList.addAll(reporterModel.getReports().stream().map(report -> report.getValue("reportSeverity").getValue().toString()).distinct().toList());
```
Comment (flomillot) [inline src/main/java/org/gridsuite/report/server/ReportService.java:149]: Please refactor to remove duplicate lines

**Review by flomillot — CHANGES_REQUESTED** [review]
Please update the swagger doc of the endpoints

**Review by dbraquart — CHANGES_REQUESTED** [review]
As discussed, severity list must not be calculated each time by reading the reports tree. We have to save it at insertion time, and then use it into the Front.
No REST api change is required.

File: src/test/java/org/gridsuite/report/server/ReportControllerTest.java [hunk 4]
```diff
@@ -308,7 +308,7 @@ public void testDeleteSubreports() throws Exception {
         insertReport(REPORT_UUID, testReportLoadflow);
 
         // Expect 4 batched inserts only and no updates
```
Comment (flomillot) [inline src/test/java/org/gridsuite/report/server/ReportControllerTest.java:310]:
```suggestion
        // Expect 5 batched inserts only and no updates
```

File: src/main/java/org/gridsuite/report/server/ReportService.java [hunk 5]
```diff
@@ -203,9 +203,13 @@ private void saveAllReportElements(ReportEntity persistedReport, ReporterModel r
     private void traverseReportModel(ReportEntity persistedReport, ReporterModel reporterModel, TreeReportEntity parentNode, List<ReportElementEntity> reportElementEntities) {
         Map<String, String> dict = new HashMap<>();
         dict.put(reporterModel.getTaskKey(), reporterModel.getDefaultName());
-        TreeReportEntity treeReportEntity = treeReportRepository.save(new TreeReportEntity(null, reporterModel.getTaskKey(), persistedReport,
+        TreeReportEntity treeReportEntity = new TreeReportEntity(null, reporterModel.getTaskKey(), persistedReport,
                 toValueEntityList(reporterModel.getTaskValues()), parentNode, dict,
-                System.nanoTime() - NANOS_FROM_EPOCH_TO_START));
+                System.nanoTime() - NANOS_FROM_EPOCH_TO_START);
+
+        treeReportEntity.getValues().add(new ReportValueEmbeddable("severityList", severityList(reporterModel), TypedValue.SEVERITY));
```
Comment (flomillot) [inline src/main/java/org/gridsuite/report/server/ReportService.java:210]: Can you do everything in the constructor ?

**Review by flomillot — APPROVED** [review]

**Comment by flomillot** [comment]
Code OK

**Review by dbraquart — CHANGES_REQUESTED** [review]
code: ok
tests: ok, but we have to discuss the compatibility, how we treat existing logs

**Review by dbraquart — APPROVED** [review]
code: ok
tests: ok

Finally we dont plan any logs migration (but node invalidation). Then we can merge as is.

**Pull request merged**